packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java

```diff
@@ -10,98 +10,59 @@
-
   private final TextureRegistry.SurfaceTextureEntry textureEntry;
-
   private final VideoPlayerCallbacks videoPlayerEvents;
-
-  private static final String USER_AGENT = "User-Agent";
-
   private final VideoPlayerOptions options;
 
-  private final DefaultHttpDataSource.Factory httpDataSourceFactory;
-
-  VideoPlayer(
+  /**
+   * Creates a video player.
+   *
+   * @param context application context.
+   * @param events event callbacks.
+   * @param textureEntry texture to render to.
+   * @param asset asset to play.
+   * @param options options for playback.
+   * @return a video player instance.
+   */
+  @NonNull
+  static VideoPlayer create(
       Context context,
       VideoPlayerCallbacks events,
       TextureRegistry.SurfaceTextureEntry textureEntry,
-      String dataSource,
-      String formatHint,
-      @NonNull Map<String, String> httpHeaders,
+      VideoAsset asset,
       VideoPlayerOptions options) {
-    this.videoPlayerEvents = events;
-    this.textureEntry = textureEntry;
-    this.options = options;
-
-    MediaItem mediaItem =
-        new MediaItem.Builder()
-            .setUri(dataSource)
-            .setMimeType(mimeFromFormatHint(formatHint))
-            .build();
-
-    httpDataSourceFactory = new DefaultHttpDataSource.Factory();
-    configureHttpDataSourceFactory(httpHeaders);
-
-    ExoPlayer exoPlayer = buildExoPlayer(context, httpDataSourceFactory);
-
-    exoPlayer.setMediaItem(mediaItem);
-    exoPlayer.prepare();
-
-    setUpVideoPlayer(exoPlayer);
+    ExoPlayer.Builder builder =
+        new ExoPlayer.Builder(context).setMediaSourceFactory(asset.getMediaSourceFactory(context));
+    return new VideoPlayer(builder, events, textureEntry, asset.getMediaItem(), options);
   }
 
-  // Constructor used to directly test members of this class.
   @VisibleForTesting
   VideoPlayer(
-      ExoPlayer exoPlayer,
+      ExoPlayer.Builder builder,
       VideoPlayerCallbacks events,
       TextureRegistry.SurfaceTextureEntry textureEntry,
-      VideoPlayerOptions options,
-      DefaultHttpDataSource.Factory httpDataSourceFactory) {
+      MediaItem mediaItem,
+      VideoPlayerOptions options) {
     this.videoPlayerEvents = events;
     this.textureEntry = textureEntry;
     this.options = options;
-    this.httpDataSourceFactory = httpDataSourceFactory;
 
-    setUpVideoPlayer(exoPlayer);
-  }
+    ExoPlayer exoPlayer = builder.build();
+    exoPlayer.setMediaItem(mediaItem);
+    exoPlayer.prepare();
 
-  @VisibleForTesting
-  public void configureHttpDataSourceFactory(@NonNull Map<String, String> httpHeaders) {
-    final boolean httpHeadersNotEmpty = !httpHeaders.isEmpty();
-    final String userAgent =
-        httpHeadersNotEmpty && httpHeaders.containsKey(USER_AGENT)
-            ? httpHeaders.get(USER_AGENT)
-            : "ExoPlayer";
-
-    unstableUpdateDataSourceFactory(
-        httpDataSourceFactory, httpHeaders, userAgent, httpHeadersNotEmpty);
+    setUpVideoPlayer(exoPlayer);
```

Comment:
not worth blocking anything but if you agree and want to change:

Seems a bit odd to me to do half the setup of `exoPlayer` out of `setUpVideoPlayer` and half of it in.

Reply:
Agreed. This will change (for the better) in the next PR.